processing/src/main/java/io/druid/query/groupby/epinephelinae/RowBasedGrouperHelper.java

```diff
@@ -0,0 +1,605 @@
+import com.metamx.common.Pair;
+import com.metamx.common.guava.Accumulator;
+import io.druid.data.input.MapBasedRow;
+import io.druid.data.input.Row;
+import io.druid.granularity.AllGranularity;
+import io.druid.granularity.QueryGranularity;
+import io.druid.query.aggregation.AggregatorFactory;
+import io.druid.query.dimension.DimensionSpec;
+import io.druid.query.extraction.ExtractionFn;
+import io.druid.query.groupby.GroupByQuery;
+import io.druid.query.groupby.GroupByQueryConfig;
+import io.druid.segment.ColumnSelectorFactory;
+import io.druid.segment.DimensionSelector;
+import io.druid.segment.FloatColumnSelector;
+import io.druid.segment.LongColumnSelector;
+import io.druid.segment.ObjectColumnSelector;
+import io.druid.segment.column.Column;
+import io.druid.segment.column.ColumnCapabilities;
+import io.druid.segment.data.IndexedInts;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+
+// this class contains shared code between GroupByMergingQueryRunnerV2 and GroupByRowProcessor
+public class RowBasedGrouperHelper
+{
+
+  public static Pair<Grouper<RowBasedKey>, Accumulator<Grouper<RowBasedKey>, Row>> createGrouperAccumulatorPair(
+      final GroupByQuery query,
+      final boolean isInputRaw,
+      final GroupByQueryConfig config,
+      final ByteBuffer buffer,
+      final int concurrencyHint,
+      final LimitedTemporaryStorage temporaryStorage,
+      final ObjectMapper spillMapper,
+      final AggregatorFactory[] aggregatorFactories
+  )
+  {
+    final GroupByQueryConfig querySpecificConfig = config.withOverrides(query);
+    final Grouper.KeySerdeFactory<RowBasedKey> keySerdeFactory = new RowBasedKeySerdeFactory(
+        query.getDimensions().size(),
+        querySpecificConfig.getMaxMergingDictionarySize() / (concurrencyHint == -1 ? 1 : concurrencyHint)
+    );
+    final RowBasedColumnSelectorFactory columnSelectorFactory = new RowBasedColumnSelectorFactory();
+    final Grouper<RowBasedKey> grouper;
+    if (concurrencyHint == -1) {
+      grouper = new SpillingGrouper<>(
+          buffer,
+          keySerdeFactory,
+          columnSelectorFactory,
+          aggregatorFactories,
+          temporaryStorage,
+          spillMapper,
+          querySpecificConfig.getBufferGrouperMaxSize(),
+          querySpecificConfig.getBufferGrouperInitialBuckets()
+      );
+    } else {
+      grouper = new ConcurrentGrouper<>(
+          buffer,
+          concurrencyHint,
+          temporaryStorage,
+          spillMapper,
+          querySpecificConfig.getBufferGrouperMaxSize(),
+          querySpecificConfig.getBufferGrouperInitialBuckets(),
+          keySerdeFactory,
+          columnSelectorFactory,
+          aggregatorFactories
+      );
+    }
+
+    final DimensionSelector[] dimensionSelectors = new DimensionSelector[query.getDimensions().size()];
```

Comment:
These are only used if `isInputRaw` is true; might as well leave this null and skip initializing them otherwise.